File: app/src/main/java/com/ternaryop/photoshelf/adapter/PhotoViewHolder.java
package com.ternaryop.photoshelf.adapter;

import java.util.List;

import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.ternaryop.lazyimageloader.ImageLoader;
import com.ternaryop.photoshelf.R;
import com.ternaryop.tumblr.TumblrAltSize;
import com.ternaryop.utils.StringUtils;
import org.joda.time.format.DateTimeFormat;

import static com.ternaryop.photoshelf.adapter.PostStyle.*;

/**
 * Created by dave on 13/04/16.
 * The ViewHolder used by Photo objects
 */
class PhotoViewHolder extends RecyclerView.ViewHolder {
    final TextView timeDesc;
    final TextView caption;
    final ImageView thumbImage;
    final ImageView menu;
    final TextView noteCountText;
    final ViewGroup tagsContainer;
    private PhotoShelfPost post;

    public PhotoViewHolder(View itemView) {
        super(itemView);
        timeDesc = itemView.findViewById(R.id.time_desc);
        caption = itemView.findViewById(R.id.caption);
        menu = itemView.findViewById(R.id.menu);
        noteCountText = itemView.findViewById(R.id.note_count);

        thumbImage = itemView.findViewById(R.id.thumbnail_image);
        tagsContainer = itemView.findViewById(R.id.tags_container);
    }

    public void bindModel(PhotoShelfPost post, ImageLoader imageLoader, int thumbnailWidth, boolean showUploadTime) {
        this.post = post;
        updateTitles(showUploadTime);
        displayImage(imageLoader, thumbnailWidth);
        setupTags();
        updateItemColors();
    }

    private void setupTags() {
        final List<String> tags = post.getTags();
        int tagsCount = tags.size();
        int viewCount = tagsContainer.getChildCount();
        int delta = tagsCount - viewCount;

        if (delta < 0) {
            for (int i = tagsCount; i < viewCount; i++) {
                tagsContainer.getChildAt(i).setVisibility(View.GONE);
            }
        } else if (delta > 0) {
            for (int i = 0; i < delta; i++) {
                tagsContainer.addView(LayoutInflater.from(tagsContainer.getContext()).inflate(R.layout.other_tag, null));
            }
        }
        for (int i = 0; i < tags.size(); i++) {
            String tag = tags.get(i);
            final TextView view = (TextView) tagsContainer.getChildAt(i);
            view.setId(R.id.tag_text_view);
            view.setText(String.format("#%s", tag));
            view.setTag(tag);
            view.setVisibility(View.VISIBLE);
        }
    }

    @SuppressWarnings("ResourceType")
    private void setColors(int resArray) {
        TypedArray array = itemView.getContext().getResources().obtainTypedArray(resArray);
        itemView.setBackground(array.getDrawable(INDEX_VIEW_BACKGROUND));
        timeDesc.setTextColor(array.getColorStateList(INDEX_TIME_DESC_TEXT_COLOR));
        caption.setTextColor(array.getColorStateList(INDEX_CAPTION_TEXT_COLOR));
        menu.setImageTintList(array.getColorStateList(INDEX_MENU_OVERFLOW_COLOR));
        noteCountText.setTextColor(array.getColorStateList(INDEX_CAPTION_TEXT_COLOR));

        final ColorStateList titleTextColor = array.getColorStateList(INDEX_TITLE_TEXT_COLOR);
        for (int i = 0; i < tagsContainer.getChildCount(); i++) {
            final TextView view = (TextView) tagsContainer.getChildAt(i);
            view.setTextColor(titleTextColor);
        }

        array.recycle();
    }

    private void displayImage(ImageLoader imageLoader, int thumbnailWidth) {
        TumblrAltSize altSize = post.getClosestPhotoByWidth(thumbnailWidth);
        setImageDimension(altSize, thumbnailWidth);

        imageLoader.displayImage(altSize.getUrl(), thumbImage);
    }

    private void setImageDimension(TumblrAltSize altSize, int thumbnailWidth) {
        ViewGroup.LayoutParams imageLayoutParams = thumbImage.getLayoutParams();
        int minThumbnailWidth = Math.max(thumbnailWidth, altSize.getWidth());
        // convert from pixel to DIP
        imageLayoutParams.width = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, minThumbnailWidth, itemView.getContext().getResources().getDisplayMetrics());
        imageLayoutParams.height = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, altSize.getHeight(), itemView.getContext().getResources().getDisplayMetrics());
    }

    private void updateTitles(boolean showUploadTime) {
        caption.setText(Html.fromHtml(StringUtils.stripHtmlTags("a|img|p|br", post.getCaption())));
        timeDesc.setText(post.getLastPublishedTimestampAsString());
        // use noteCountText for both uploadTime and notes
        if (showUploadTime) {
            showUploadTime();
        } else {
            updateNote();
        }
    }

    private void showUploadTime() {
        noteCountText.setVisibility(View.VISIBLE);
        noteCountText.setText(itemView.getResources().getString(R.string.uploaded_at_time, DateTimeFormat.forPattern("dd/MM/yyyy HH:mm").print(post.getTimestamp() * 1000)));
    }

    private void updateNote() {
        int noteCount = (int) post.getNoteCount();
        if (noteCount > 0) {
            noteCountText.setVisibility(View.VISIBLE);
            noteCountText.setText(itemView.getContext().getResources().getQuantityString(
                    R.plurals.note_title,
                    noteCount,
                    noteCount));
        } else {
            noteCountText.setVisibility(View.GONE);
        }
    }

    public void setOnClickListeners(View.OnClickListener listener) {
        if (listener != null) {
            setTagsClickListener(listener);
            final int position = getAdapterPosition();
            thumbImage.setOnClickListener(listener);
            thumbImage.setTag(position);

            menu.setOnClickListener(listener);
            menu.setTag(position);
        }
    }

    public void setOnClickMultiChoiceListeners(View.OnClickListener listener, View.OnLongClickListener longClickListener) {
        if (listener != null) {
            itemView.setOnClickListener(listener);
            itemView.setOnLongClickListener(longClickListener);
            itemView.setLongClickable(true);
            itemView.setTag(getAdapterPosition());
        }
    }

    private void updateItemColors() {
        switch (post.getScheduleTimeType()) {
            case POST_PUBLISH_NEVER:
                setColors(R.array.post_never);
                break;
            case POST_PUBLISH_FUTURE:
                setColors(R.array.post_future);
                break;
            default:
                setColors(post.getGroupId() % 2 == 0 ? R.array.post_even : R.array.post_odd);
                break;
        }
    }

    private void setTagsClickListener(View.OnClickListener listener) {
        tagsContainer.setTag(getAdapterPosition());
        for (int i = 0; i < tagsContainer.getChildCount(); i++) {
            tagsContainer.getChildAt(i).setOnClickListener(listener);
        }
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/adapter/photo/PhotoViewHolder.kt
package com.ternaryop.photoshelf.adapter.photo

import android.annotation.SuppressLint
import android.support.v7.widget.RecyclerView
import android.text.format.DateUtils.SECOND_IN_MILLIS
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.ternaryop.lazyimageloader.ImageLoader
import com.ternaryop.photoshelf.R
import com.ternaryop.photoshelf.adapter.POST_STYLE_INDEX_CAPTION_TEXT_COLOR
import com.ternaryop.photoshelf.adapter.POST_STYLE_INDEX_MENU_OVERFLOW_COLOR
import com.ternaryop.photoshelf.adapter.POST_STYLE_INDEX_TIME_DESC_TEXT_COLOR
import com.ternaryop.photoshelf.adapter.POST_STYLE_INDEX_TITLE_TEXT_COLOR
import com.ternaryop.photoshelf.adapter.POST_STYLE_INDEX_VIEW_BACKGROUND
import com.ternaryop.photoshelf.adapter.PhotoShelfPost
import com.ternaryop.photoshelf.util.text.fromHtml
import com.ternaryop.tumblr.TumblrAltSize
import com.ternaryop.utils.StringUtils
import org.joda.time.format.DateTimeFormat

/**
 * Created by dave on 13/04/16.
 * The ViewHolder used by Photo objects
 */
class PhotoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    private val timeDesc: TextView = itemView.findViewById(R.id.time_desc)
    private val caption: TextView = itemView.findViewById(R.id.caption)
    private val thumbImage: ImageView = itemView.findViewById(R.id.thumbnail_image)
    private val menu: ImageView = itemView.findViewById(R.id.menu)
    private val noteCountText: TextView = itemView.findViewById(R.id.note_count)
    private val tagsContainer: ViewGroup = itemView.findViewById(R.id.tags_container)
    private lateinit var post: PhotoShelfPost

    fun bindModel(post: PhotoShelfPost, imageLoader: ImageLoader, thumbnailWidth: Int, showUploadTime: Boolean) {
        this.post = post
        updateTitles(showUploadTime)
        displayImage(imageLoader, thumbnailWidth)
        setupTags()
        updateItemColors()
    }

    @SuppressLint("InflateParams")
    private fun setupTags() {
        val tags = post.tags
        val tagsCount = tags.size
        val viewCount = tagsContainer.childCount
        val delta = tagsCount - viewCount

        if (delta < 0) {
            for (i in tagsCount until viewCount) {
                tagsContainer.getChildAt(i).visibility = View.GONE
            }
        } else if (delta > 0) {
            for (i in 0 until delta) {
                tagsContainer.addView(LayoutInflater.from(tagsContainer.context).inflate(R.layout.other_tag, null))
            }
        }
        for (i in tags.indices) {
            val tag = tags[i]
            val view = tagsContainer.getChildAt(i) as TextView
            view.id = R.id.tag_text_view
            view.text = String.format("#%s", tag)
            view.tag = tag
            view.visibility = View.VISIBLE
        }
    }

    private fun setColors(resArray: Int) {
        val array = itemView.context.resources.obtainTypedArray(resArray)
        itemView.background = array.getDrawable(POST_STYLE_INDEX_VIEW_BACKGROUND)
        timeDesc.setTextColor(array.getColorStateList(POST_STYLE_INDEX_TIME_DESC_TEXT_COLOR))
        caption.setTextColor(array.getColorStateList(POST_STYLE_INDEX_CAPTION_TEXT_COLOR))
        menu.imageTintList = array.getColorStateList(POST_STYLE_INDEX_MENU_OVERFLOW_COLOR)
        noteCountText.setTextColor(array.getColorStateList(POST_STYLE_INDEX_CAPTION_TEXT_COLOR))

        val titleTextColor = array.getColorStateList(POST_STYLE_INDEX_TITLE_TEXT_COLOR)
        for (i in 0 until tagsContainer.childCount) {
            val view = tagsContainer.getChildAt(i) as TextView
            view.setTextColor(titleTextColor)
        }

        array.recycle()
    }

    private fun displayImage(imageLoader: ImageLoader, thumbnailWidth: Int) {
        val altSize = post.getClosestPhotoByWidth(thumbnailWidth) ?: return
        setImageDimension(altSize, thumbnailWidth)

        imageLoader.displayImage(altSize.url, thumbImage)
    }

    private fun setImageDimension(altSize: TumblrAltSize, thumbnailWidth: Int) {
        val minThumbnailWidth = Math.max(thumbnailWidth, altSize.width)
        // convert from pixel to DIP
        with (thumbImage.layoutParams) {
            width = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, minThumbnailWidth.toFloat(), itemView.context.resources.displayMetrics).toInt()
            height = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, altSize.height.toFloat(), itemView.context.resources.displayMetrics).toInt()
        }
    }

    private fun updateTitles(showUploadTime: Boolean) {
        caption.text = StringUtils.stripHtmlTags("a|img|p|br", post.caption).fromHtml()
        timeDesc.text = post.lastPublishedTimestampAsString
        // use noteCountText for both uploadTime and notes
        if (showUploadTime) {
            showUploadTime()
        } else {
            updateNote()
        }
    }

    private fun showUploadTime() {
        noteCountText.visibility = View.VISIBLE
        noteCountText.text = itemView.resources.getString(R.string.uploaded_at_time, DateTimeFormat.forPattern("dd/MM/yyyy HH:mm").print(post.timestamp * SECOND_IN_MILLIS))
    }

    private fun updateNote() {
        val noteCount = post.noteCount.toInt()
        if (noteCount > 0) {
            noteCountText.visibility = View.VISIBLE
            noteCountText.text = itemView.context.resources.getQuantityString(
                    R.plurals.note_title,
                    noteCount,
                    noteCount)
        } else {
            noteCountText.visibility = View.GONE
        }
    }

    fun setOnClickListeners(listener: View.OnClickListener?) {
        if (listener != null) {
            setTagsClickListener(listener)
            val position = adapterPosition
            thumbImage.setOnClickListener(listener)
            thumbImage.tag = position

            menu.setOnClickListener(listener)
            menu.tag = position
        }
    }

    fun setOnClickMultiChoiceListeners(listener: View.OnClickListener?, longClickListener: View.OnLongClickListener) {
        if (listener != null) {
            itemView.setOnClickListener(listener)
            itemView.setOnLongClickListener(longClickListener)
            itemView.isLongClickable = true
            itemView.tag = adapterPosition
        }
    }

    private fun updateItemColors() {
        when (post.scheduleTimeType) {
            PhotoShelfPost.ScheduleTime.POST_PUBLISH_NEVER -> setColors(R.array.post_never)
            PhotoShelfPost.ScheduleTime.POST_PUBLISH_FUTURE -> setColors(R.array.post_future)
            else -> setColors(if (post.groupId % 2 == 0) R.array.post_even else R.array.post_odd)
        }
    }

    private fun setTagsClickListener(listener: View.OnClickListener) {
        tagsContainer.tag = adapterPosition
        for (i in 0 until tagsContainer.childCount) {
            tagsContainer.getChildAt(i).setOnClickListener(listener)
        }
    }
}
